[WIP] Keybinding Overhaul

calmilamsy opened the pull request:
<!-- The notes within these arrows are for you but can be deleted. -->

## Summary

**DEPENDS ON #797**
- Add the ability to remap hotkeys.
- Adds in a dynamically generated keybind menu, replacing the old one.

## Pictures/Videos

(First prototype) 
https://cdn.discordapp.com/attachments/603622883075293214/879812051550216283/2021-08-24_20-36-35.mp4

## Changes to Files

- Deletes a bunch of things that are no longer required from the ServerLobby scene.
- Adds in a bunch of logic into GeneralSettingsUIHelper.cs to generate the new keybind menu.
- Changes the BindingButton prefab to have a script attached that rebinds the contained key.
- Changes every script that is ***CURRENTLY*** used in game that I could find that handles keypresses to use the new input system.

## Technical Notes

This one will be a long one once I am done, and I will detail way more once what people should do for keybinds is more set in stone.

- Uses the new Unity Input System (1.1.0-pre.6, requires #797).
- Adds in a helpful public static class `SS3D.Input.InputHelper` that contains all the keybinds which can be accessed from anywhere.
- Adds the ability to dynamically add and remove hotkeys really easily. The keybindings UI currently does NOT support these yet though.

<!-- Provide a more technical description of your changes to help save the reviewers some time. -->

## Known Issues

- Rebinding the scroll wheel for zooming in and out using the keybinds UI is not supported.
- Might be an idea to provide a bit of visual feedback to the user about what key they are rebinding.
- Renaming or removing controls in the InputActions object will cause the binding overrides to implode and throw errors. I have resorted to just letting it stay on default bindings if it fails to load the bindings, overwriting the old bindings config with any new bindings made.

**All of this can change as I make more progress, this is just so folk know I am working on it, and can provide feedback.**

calmilamsy committed:
Add pressing "T" to open and focus chat

calmilamsy committed:
Actually commit InputManager

calmilamsy committed:
Add gitignore for .vsconfig

calmilamsy committed:
First prototype

Known issues:
Keys are not rebound properly, and will fail to rebind again once rebound once.
API and namespaces are completely unfinished.

calmilamsy committed:
Super incorrect usage (I think) of the new input system

Problem is that I have zero documented alternatives to EventSystem.current.IsPointerOverGameObject()

Also Unity doesn't document jack shit about it's new input system for our use case.

calmilamsy committed:
Cursed code 2: electric boogaloo

This does not work, do not look at this code, I want to forget I even tried this.

calmilamsy committed:
Cursed code 3: lightning tree

Keybinding is done, but I need to do this in a much better way.
Also like half the code still uses the old system.

Ryan089 committed:
Updated Unity to version 2019.4.29f1 (2019 LTS)

Ryan089 committed:
Removed unnecessary (?) files from commit.

calmilamsy committed:
Merge branch 'pr/797' into new-inputsystem

calmilamsy committed:
Potentially usable code

calmilamsy committed:
Merge branch 'new-inputsystem' into keybinding-overhaul

calmilamsy committed:
Fix examine UI

calmilamsy committed:
Add some comments, rename some binds, and ditch the use of vector inputs where possible

calmilamsy committed:
Rename InputHelper.inp to InputHelper.inputs and rename the Run control to Toggle Run

calmilamsy committed:
Re-add loading binding overrides, discarding if they error due to a stupid unity thing

calmilamsy committed:
Create ADDING_CONTROLS.md

calmilamsy committed:
Make binding button debug output more descriptive

calmilamsy committed:
Merge remote-tracking branch 'upstream/master' into keybinding-overhaul

cosmiccoincidence commented on .gitignore:
```diff
@@ -29,6 +29,9 @@
 .vscode/
 
+# Visual studio project config
+.vsconfig
```
I just added this in a commit on one of Ryan's PRs that was just merged. You can remove this file from your PR.

calmilamsy replied:
Removed in latest push.

Ryan089 commented:
This looks really cool. Thanks for the effort you've put into it. Just wanted to let you know it'll probably take quite a while to finish the review, but rest assured it is being looked at!

cosmiccoincidence commented:
> Rebinding the scroll wheel for zooming in and out using the keybinds UI is not supported.

This is fine because the controls for the camera we have will not be available to players in our final design. These camera controls will be limited to admins only (along with some other controls). But we don't really have that distinguished as of yet.

> Might be an idea to provide a bit of visual feedback to the user about what key they are rebinding.

Did you have an idea in mind on how to do this?

cosmiccoincidence commented on Documents/ADDING_CONTROLS.md:
```diff
@@ -0,0 +1,26 @@
+
+## Technical Notes
+
+- If you set your action to hold, triggered and acts in the exact same way as if you set the press type to Press and Release.
```
Typo on this line?

calmilamsy replied:
Yup, added an extra `and`, woops.

cosmiccoincidence reviewed (commented):
Some more commenting of what certain code or files do would be appreciated. This will tremendously help new developers in the future get a quicker grasp of what's going on.
A few files I noticed that could use a summary:
- InputHelper.cs
- KeybindingHelper.cs
- KeybindingManager.cs

calmilamsy committed:
Fix typo in CONTROLS.md

calmilamsy committed:
Merge remote-tracking branch 'upstream/master' into keybinding-overhaul

calmilamsy committed:
Merge branch 'master' into keybinding-overhaul

calmilamsy committed:
Remove outdated gitignore entry

calmilamsy commented:
Will work on the code comments and button highlighting on pressing the rebind button tomorrow.

calmilamsy committed:
Remove old input files

No longer needed with the new system.

calmilamsy committed:
Make the rebind button visually respond to being hovered over and clicked

calmilamsy committed:
Clean up code on BindingButtonHelper, fix name in InputHelper

Apparently statics in C# use camelcase. Why?
Also added summaries to everything I added.
Made the rebind button say "Rebinding" when clicked to provide more feedback to the user.

calmilamsy commented:
> 
> 
> Some more commenting of what certain code or files do would be appreciated. This will tremendously help new developers in the future get a quicker grasp of what's going on.
> A few files I noticed that could use a summary:
> 
>     * InputHelper.cs
> 
>     * KeybindingHelper.cs
> 
>     * KeybindingManager.cs

KeybindingHelper and KeybindingManager were holdovers from the old system. Removed them.
I have also added summaries to everything I added, I think.

cosmiccoincidence reviewed (commented):
- Are these the defaults? Some of these do not seem correct if they are meant to be the defaults.
![image](https://user-images.githubusercontent.com/41941008/134225327-34bf0cca-ef71-41bc-a3d5-b043ef944d01.png)

- I'm a bit confused by this section here. Not sure if we should allow rebinding ESC (the exit key).
![image](https://user-images.githubusercontent.com/41941008/134225584-234d7e15-4f56-40ac-8dff-b4922adf9902.png)

- There is a scroll bar in the settings menu, but I need to use my mouse to use it. I should also be able to use my scroll wheel on my mouse for this.

- I can't remember, were you going to add a way to reset controls to the default in this PR or are you waiting to add that in another PR?

calmilamsy commented:
Pretty sure I coped the defaults from the old settings,  I will double check just in case.
I can blacklist escape from being rebound.
Mouse wheel worked in testing, unsure why that isnt working for you.
And I am fairly certain I said I would add resetting to default in another PR, cause it would require rewriting and adding a ton of code that isnt really in the scope of this PR.

cosmiccoincidence commented:
I recloned the branch tested it all again and I'm still getting weird default keybinds and the scroll bar not working with the mouse wheel.

Also don't forget to disable the ability to rebind the misc category.

Lastly, something new I noticed is that the intent button recently got an update to its appearance but that doesn't seem to be present in your branch here. Consider merging in master to amend that.

calmilamsy committed:
Fix default bindings and mouse wheel scrolling in UIs

calmilamsy committed:
Merge remote-tracking branch 'upstream/master' into keybinding-overhaul

calmilamsy committed:
Make #805 use new input system

calmilamsy committed:
Hide submit binding

Also fixes an error and incorrect binding display.

calmilamsy committed:
Rename Misc binding group to System and make it unrebindable

calmilamsy commented:
I *think* I have done all you requested. Please double check to make sure.

calmilamsy commented:
Moved back to WIP cause I am changing almost everything about the Keybind UI.

cosmiccoincidence commented:
Closing this for now and marking it Derelict so we come back to it.
